Special topic chapter for finalizers and weak references (#1265)
mmtk/mmtk-core: docs/userguide/src/portingguide/concerns/weakref.md

    +   Do this with `ObjectReference::is_reachable()`.
-   **Query forwarded address**: If an object has already been reached, the VM binding can further
    query the new address of an object.  This is needed to support copying GC.
    +   Do this with `ObjectReference::get_forwarded_object()`.
-   **Retain objects**: If an object has not been reached at this time, the VM binding can
    optionally demand the object to be retained.  That object *and all descendants* will be kept
    alive during this GC.
    +   Do this with the `tracer_context` argument of `process_weak_refs`.
-   **Request another invocation**: The VM binding can request `Scanning::process_weak_refs` to be
    called again after computing the transitive closure that includes *retained objects and their
    descendants*.  This helps handling multiple levels of weak reference strength.
    +   Do this by returning `true` from `process_weak_refs`.

The `Scanning::process_weak_refs` function also gives the VM binding a chance to perform other
operations, including (but not limited to)

-   **Do clean-up operations**: The VM binding can perform clean-up operations, or queue them to be
    executed after GC.

> -   **update fields** that contain weak references.

Reviewer: ```suggestion
-   **Update fields** that contain weak references.
```

    -   **Forward the field**: It can write the forwarded address of the referent if moved by a
        copying GC.
    -   **Clear the field**: It can clear the field if the referent has not been reached and the
        binding decides it is unreachable.

Using those primitive operations, the VM binding can support different flavors of finalizers and/or
weak references.  We will discuss common use cases in the following sections.

## Supporting finalizers

> Different VMs define "finalizer" differently, but they all involve performing operations when an

Reviewer: Technically it's the language specification (if it exists) that defines finalization, no? Same for other locations below.

Author: This is a bit subtle.  Some languages (Java for example) define the semantics of finalizers clearly.  For other languages or VMs, it is a bit more subtle.

For Ruby, finalizer is not part of the language.  There are two kinds of finalizers, both are VM details.
1.  `ObjectSpace.define_finalizer`: The entire `ObjectSpace` module is explicitly documented as implementation-specific and not portable.
2.  The `obj_free` function.  It is for fundamental and internal types in CRuby and its C extensions.  It is not part of the language, either.

By the way, the idiomatic way to close resources in Ruby is enclosing the resource in a block, for example `File.open(...) {|f| use(f) }`.

And even for the same language, different VM implementations may implement it slightly differently.  OpenJDK and JikesRVM are, for example, different.

And for Java, although the language spec defines finalization, the semantics of weak references is defined in neither the language spec nor the VM spec, but the API.  I still consider the API as part of the language, and the VM still needs to support the Java API anyway, including its weak reference semantics.

But I can rephrase this part.  I'll say different *languages and* VMs define finalizer differently.

unreachable, do something about them.

### Identifying finalizable objects

Some VMs determine whether an object is finalizable by its type.  In Java, for example, an object is
finalizable if its `finalize()` method is overridden.  The VM binding can maintain a list of
finalizable objects, and register instances of such types into that list when they are constructed.

Some VMs can dynamically attach finalizing operations to individual objects after objects are
created.  The VM binding can maintain a list of objects with attached finalizers, or maintain a
(weak) hash map that maps finalizable objects to its associated finalizers.

### When to run finalizers?

Depending on the finalizer semantics in different VMs, finalizers can be executed during GC or
during mutator time after GC.

The VM binding can run finalizers immediately in `Scanning::process_weak_refs` when finding a
finalizable object unreachable.  Beware that executing finalizers can be time-consuming.  The VM
binding can creating work packets and let each work packet process a part of all finalizable
objects.  In this way, multiple GC workers can process finalizable objects in parallel.  The

> `Scanning::process_weak_refs` function is executed in the `VMRefClosure` stage, so the created work

Reviewer: Should we be using the term "stage" for work buckets?

Author: That's our current terminology.  We used to call it "phase" in JikesRVM.  Currently the type name is `WorkBucketStage` defined in `src/scheduler/work_bucket.rs`.

function can also move the object.

Under the hood, `tracer_context.with_tracer` creates a queue and calls the closure.  The `tracer`
implements the `ObjectTracer`  trait, and is just an interface that provides the `trace_object`
method.  Objects retained by `tracer.trace_object` will be enqueued.  After the closure returns,
`with_tracer` will split the queue into reasonably-sized work packets and add them to the
`VMRefClosure` work bucket.  Those work packets will trace the retained objects and their
descendants, effectively expanding the transitive closure to include all objects reachable from the
retained objects.  Because of the overhead of creating queues and work packets, the VM binding
should **retain as many objects as needed in one invocation of `with_tracer`, and avoid calling
`with_tracer` again and again for each object**.

**Don't do this**:

```rust
for object in objects {
    tracer_context.with_tracer(worker, |tracer| { // This is expensive! DON'T DO THIS!
        tracer.trace_object(object);
    });
}
```

Keep in mind that **tracer_context implements the `Clone` trait**.  As introduced in the *When to

> run finalizers* section, the VM binding can use work packets to parallelize finalizer processing.

Reviewer: Back link to section?

If finalizable objects need to be retained, the VM binding can clone the `trace_context` and give
each work packet a clone of `tracer_context`.

### WARNING: object resurrection

If the VM binding retains an unreachable object for finalization, and the finalizer writes a
reference of that object into a place readable by application threads, including global or static
variable, then the previously unreachable object will become reachable by the application again.
This phenomenon is known as **"resurrection"**, and can be surprising to the programmers.

> Developers of VM bindings of existing VMs may have no choice but implementing the finalizer

Reviewer: ```suggestion
Developers of VM bindings of existing VMs may have no choice but to implement the finalizer
```

need to reconstruct the list in every GC, while discovering during tracing can avoid visiting dead
objects.  Depending on the nature of your VM, one method may be easier to implement than the other,
especially if your VM's existing GC has already implemented weak reference processing in some way.

### Associated clean-up operations

Some languages and VMs allow certain clean-up operations to be associated with weak references, and
will be executed after the weak reference is cleared.

Such clean-up operations can be supported similar to finalizers.  While we enumerate weak references
in `Scanning::process_weak_refs`, we clear weak references to unreachable objects.  Depending on the
semantics, we may choose to execute the clean-up operations immediately, or enqueue them to be
executed after GC.  We may retain the unreachable referent if we need to.

### Soft references

Java has a special kind of weak reference: `SoftReference`.  The API allows the GC to choose between
(1) retaining softly reachable referents, and (2) clearing references to softly reachable objects.
When using MMTk, there are two ways to implement this semantics.

> The easiest way is **treating `SoftReference` as strong references in non-emergency GCs, and

Reviewer: Perhaps add TODO to glossary for "emergency GC"

supporting multiple levels of reference strengths, which will be introduced in the next section.

### Multiple levels of reference strength

Some VMs support multiple levels of weak reference strengths.  Java, for example, has
`SoftReference`, `WeakReference`, `FinalizerReference` (internal) and `PhantomReference`, in the
order of decreasing strength.

This can be supported by running `Scanning::process_weak_refs` multiple times.  If
`process_weak_refs` returns `true`, it will be called again after all pending work packets in the
`VMRefClosure` stage has been executed.  Those pending work packets include all work packets that
compute the transitive closure from objects retained during `process_weak_refs`.  This allows the VM
binding to expand the transitive closure multiple times, each handling weak references at different
levels of strength.

Take Java as an example,  we may run `process_weak_refs` four times.

1.  Visit all `SoftReference`.
    -   If the referent has been reached, then
        -   forward the referent field.
    -   If the referent has not been reached, yet, then

>         -   if it is not emergency GC, then

Reviewer: ```suggestion
        -   if it is not an emergency GC, then
```

            -   retain the referent and update the referent field.

>         -   it it is emergency GC, then

Reviewer: ```suggestion
        -   it it is an emergency GC, then
```

            -   clear the referent field,
            -   remove the `SoftReference` from the list of soft references, and
            -   optionally enqueue it to the associated `ReferenceQueue` if it has one.

>     -   (This step may expand the transitive closure in emergency GC if any referents are retained.)

Reviewer: ```suggestion
    -   (This step may expand the transitive closure in an emergency GC if any referents are retained.)
```

        }
        State::ProcessFinalizableObjects => {
            process_finalizable_objects(...);
            *state = State::ProcessPhantomReferences;
            return true; // Run this function again.
        }
        State::ProcessPhantomReferences => {
            process_phantom_references(...);
            *state = State::ProcessSoftReference
            return false; // Proceed to the Release stage.
        }
    }
}
```

### Ephemerons

An [Ephemeron] has a *key* and a *value*, both of which are object references.  The key is a weak
reference, while the value keeps the referent alive only if both the ephemeron itself and the key
are reachable.

[Ephemeron]: https://dl.acm.org/doi/10.1145/263700.263733

To support ephemerons, the VM binding needs to identify ephemerons.  This includes ephemerons as
individual objects, objects that contain ephemerons, and, equivalently, objects that contain
key/value fields that have semantics similar to ephemerons.

The following is the algorithm for processing ephemerons.  It gradually discovers ephemerons as we
do the tracing.  We maintain a queue of ephemerons which is empty before the `Closure` stage.

> 1.  In `Scanning::scan_object` and `Scanning::scan_object_and_trace_edges`, we enqueue ephemerons as

Reviewer: Have to phrase this a bit delicately since there are two queues here. The ephemeron queue and the scanning/tracing queue. I got tripped up by you saying "enqueuing ephemerons but not tracing them".

Author: Yes.  This "enqueue" refers to the ephemeron queue.

last GC) as candidates of garbage, and will assume all *old objects* (i.e. objects survived the last
GC) are live.

The VM binding can query if the current GC is a nursery GC by calling

```rust
let is_nursery_gc = mmtk.get_plan().generational().is_some_and(|gen|
    gen.is_current_gc_nursery());
```

The VM binding can make use of this information when processing finalizers and weak references.  In
a minor GC,

-   The VM binding only needs to visit **finalizable objects allocated since the last GC**.  Other
    finalizable objects must be old and will not be considered dead.
-   The VM binding only needs to visit **weak reference slots written since the last GC**.  Other
    slots must be pointing to old objects (if not `null`).  For weak hash tables, if existing
    entries are immutable, it is sufficient to only visit newly added entries.

Implementation-wise, the VM binding can split the lists or hash tables into two parts: one for old
entries and another for young entries.

### Copying versus non-copying GC

MMTk provides both copying and non-copying GC plans.  `MarkSweep` never moves any objects.

> `MarkCompact`, `SemiSpace` always moves all objects.  Immix-based plans sometimes do non-copying GC,

Reviewer: Ehhh. That's not true since LOS/non-moving objects are not moved. Should say normal objects there maybe?

Author: Sure.  I'll make it clear objects in the LOS, ImmortalSpace, etc. don't move.
